src/main/scala/cloud/drdrdr/utils/AWSCredentials.scala

```diff
@@ -6,165 +6,416 @@ package cloud.drdrdr.utils
+      response.entity.dataBytes.map(_.utf8String).grouped(Int.MaxValue).runWith(Sink.head) map {
+        case responseInfo =>
+          val responseData = responseInfo.mkString
+          val responseJson = responseData.parseJson
+          var keyId: Option[String] = None
+          var accessKey: Option[String] = None
+          var token: Option[String] = None
+          var expiration: Option[String] = None
+          val jsonMap = responseJson.asJsObject().fields
+          val jsKeyId = jsonMap.get("AccessKeyId")
+          if (jsKeyId.isDefined)
+            keyId = Some(jsKeyId.get.toString() replaceAll("[\"]", ""))
+          val jsAccessKey = jsonMap.get("SecretAccessKey")
+          if (jsAccessKey.isDefined)
+            accessKey = Some(jsAccessKey.get.toString() replaceAll("[\"]", ""))
+          val jsToken = jsonMap.get("Token")
+          if (jsToken.isDefined)
+            token = Some(jsToken.get.toString() replaceAll("[\"]", ""))
+          val jsExpiration = jsonMap.get("Expiration")
+          if (jsExpiration.isDefined)
+            expiration = Some(jsExpiration.get.toString() replaceAll("[\"]", ""))
+          validCredentials(keyId, accessKey, token, expiration)
+      }
+  }
+
+  // gets the role name off the ec2 instance
+  protected def getAmazonEC2RoleName(timeout: Int = 300)
+                                    (implicit ec: ExecutionContext, s: ActorSystem, m: ActorMaterializer): Future[Option[String]] = {
+      import spray.json._
+
+      val request = HttpRequest(HttpMethods.GET, "http://169.254.169.254/latest/meta-data/iam/info")
+//      val instanceFuture = Http().singleRequest(request).flatMap { response =>
+      val instanceFuture = sendRequest(request).flatMap { response =>
+        response.entity.dataBytes
+          .fold(ByteString.empty)(_ ++ _)
+          .map(_.utf8String)
+          .runWith(Sink.head)
+          .map(_.parseJson)
+          .map {
+            _.asJsObject.fields.get("InstanceProfileArn")
+              .map(_.toString().split("/").last.replaceAll("[\"]", ""))
+          }
+      }
+      Future.firstCompletedOf(
+        List(instanceFuture, after(timeout milliseconds, s.scheduler)(Future.successful(None))))
+  }
+  /**
+   * gets the aws credentials associated with the role of the ec2 instance
+ *
+   * @param timeout time to wait for the ec2 response in miliseconds
+   * @param ec implicit execution context
+   * @param s implicit actor system
+   * @param m implicit actor materializer
+   * @return aws credentials or None
+   */
+  @throws(classOf[Exception])
+  def getAmazonEC2Credentials(timeout: Int = 300)(implicit ec: ExecutionContext, s:ActorSystem, m: ActorMaterializer): Future[Option[AWSPermissions]] = {
+    val roleName = getAmazonEC2RoleName(timeout)
+    val ec2Credentials = roleName flatMap{
+      case Some(role) =>
+        getEC2RoleCredentials(role, timeout)
+      case None =>
```

Comment:
Woops. The None case should return a Future.successful(None) so that the chain doesn't throw an exception if this method fails. The case _ is unnecessary